nemo_text_processing/text_normalization/ru/taggers/electronic.py

```diff
@@ -0,0 +1,117 @@
+        username = (
+            pynutil.insert("username: \"")
+            + NEMO_ALPHA
+            + pynini.closure(NEMO_ALPHA | NEMO_DIGIT | pynini.union(*accepted_symbols))
+            + pynutil.insert("\"")
+            + pynini.cross('@', ' ')
+        )
+        domain_graph = (
+            NEMO_ALPHA
+            + (pynini.closure(NEMO_ALPHA | NEMO_DIGIT | pynini.accep('-') | pynini.accep('.')))
+            + (NEMO_ALPHA | NEMO_DIGIT)
+        )
+        domain_graph = pynutil.insert("domain: \"") + domain_graph + pynutil.insert("\"")
+        tagger_graph = (username + domain_graph).optimize()
+
+        # verbalizer
+        graph_digit = pynini.string_file(get_abs_path("data/numbers/digits_nominative_case.tsv")).optimize()
+        graph_symbols = pynini.string_file(get_abs_path("data/electronic/symbols.tsv")).optimize()
+        user_name = (
+            pynutil.delete("username:")
+            + delete_space
+            + pynutil.delete("\"")
+            + (
+                pynini.closure(
+                    pynutil.add_weight(graph_digit + insert_space, 1.09)
+                    | pynutil.add_weight(pynini.closure(graph_symbols + pynutil.insert(" ")), 1.09)
+                    | pynutil.add_weight(NEMO_NOT_QUOTE + insert_space, 1.1)
+                )
+            )
+            + pynutil.delete("\"")
+        )
+
+        domain_default = (
+            pynini.closure(NEMO_NOT_QUOTE + insert_space)
+            + pynini.cross(".", "точка ")
+            + NEMO_NOT_QUOTE
+            + pynini.closure(insert_space + NEMO_NOT_QUOTE)
+        )
+
+        server_default = (
+            pynini.closure((graph_digit | NEMO_ALPHA) + insert_space, 1)
+            + pynini.closure(graph_symbols + insert_space)
+            + pynini.closure((graph_digit | NEMO_ALPHA) + insert_space, 1)
+        )
+        server_common = pynini.string_file(get_abs_path("data/electronic/server_name.tsv")) + insert_space
+        domain_common = pynini.cross(".", "точка ") + pynini.string_file(get_abs_path("data/electronic/domain.tsv"))
+        domain = (
+            pynutil.delete("domain:")
+            + delete_space
+            + pynutil.delete("\"")
+            + (pynutil.add_weight(server_common, 1.09) | pynutil.add_weight(server_default, 1.1))
+            + (pynutil.add_weight(domain_common, 1.09) | pynutil.add_weight(domain_default, 1.1))
+            + delete_space
+            + pynutil.delete("\"")
+        )
+
+        graph = user_name + delete_space + pynutil.insert("собака ") + delete_space + domain + delete_space
+        # replace all latin letters with their Ru verbalization
+        verbalizer_graph = (graph.optimize() @ (pynini.closure(TO_LATIN | RU_ALPHA | pynini.accep(" ")))).optimize()
+        verbalizer_graph = verbalizer_graph.optimize()
+
+        self.final_graph = (tagger_graph @ verbalizer_graph).optimize()
+        self.fst = self.add_tokens(self.final_graph).optimize()
```

Comment:
can we use tagger_graph here? if not, why separating between tagger and verbalizer? is it because of speed when permutations are done?

Reply:
what happens in sparrowhawk if you don't have class labels like "domain" etc